File: week4/lee/단속카메라.py
def solution(routes):
    answer = 0
    routes.sort(key=lambda x:x[1]) # 진출시간 기준으로 정렬

    location = -30001 # 
    for start, end in routes:
        if(location < start): # 카메라 설치해야하는 경우
            answer += 1
            location = end
    return answer

print(solution([[-20,-15], [-14,-5], [-18,-13], [-5,-3]]	))

File: week4/eunbeen/단속카메라.java
import java.util.*;

class Car implements Comparable<Car>{
    int input, output;
    
    Car(int input, int output){
        this.input = input;
        this.output = output;
    }
    
    @Override
    public int compareTo(Car car){
        return Integer.compare(this.input, car.input);
    }
    
    @Override
    public String toString(){
        return "input: " + input + ", output: " + output;
    }
}

class Solution {
    public int solution(int[][] routes) {
        int answer = 0;
        Stack<Integer> stack = new Stack(); 
        PriorityQueue<Car> queue = new PriorityQueue();
        
        for(var route: routes){
            queue.add(new Car(route[0], route[1]));
        }
        
        while(queue.size() > 0){
            var current = queue.poll();
            int camera = current.output;
            
            while(queue.size() > 0 && queue.peek().input <= camera){
                var next = queue.poll();
                camera = Math.min(next.output, camera);
            }
            answer += 1;
        }
        
        return answer;
    }
}